### PR TITLE
Add dimension comparison

### PR DESCRIPTION
**Pull request by djbarnwal**
Add dimension comparison to the dashboard 

https://www.notion.so/rilldata/Enable-users-to-compare-time-series-trends-of-different-dimension-values-in-the-Overview-page-3bb82762a902415f8dce7178038a636c

### CODE REVIEW

**Comment by ericpgreen2** [comment]
QA finding: circles are not centered with the text 
image

**Comment by ericpgreen2** [comment]
UX finding:
1. On Cloud (maybe on Local too, but haven't checked), navigate to a dashboard fresh w/o any URL state
2. Click on "Compare" button for a dimension
3. Change the time grain
4. It's unresponsive, and I see this in the browser console:

image

**Comment by ericpgreen2** [comment]
UX nit: when a dimension is actively being compared, I'd expect that clicking its "Compare" button would remove the comparison

**Comment by ericpgreen2** [comment]
Here's a Notion doc with UXQA feedback: https://www.notion.so/rilldata/Dimension-Comparison-UXQA-eb2c55dab48f49399c9b8af6780e1ce8

File: web-common/src/features/dashboards/big-number/MeasureBigNumber.svelte [hunk 1]
```diff
@@ -42,7 +42,11 @@
   $: measureIsPercentage = formatPresetEnum === FormatPreset.PERCENTAGE;
 </script>
 
-<div class="flex flex-col pl-1 {withTimeseries ? 'mt-2' : 'justify-between'}">
+<button
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/big-number/MeasureBigNumber.svelte:45]: This `button` doesn't have an `on:click`

File: web-common/src/features/dashboards/leaderboard/leaderboard-utils.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ export type LeaderboardItemData = {
   // selection is not enough to determine if the item is included
   // or excluded; for that we need to know the leaderboard's
   // include/exclude state
-  selected: boolean;
+  selected: number;
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/leaderboard/leaderboard-utils.ts:21]: Now it's `selectedIndex`?

File: web-common/src/features/dashboards/time-series/TimeSeriesChartContainer.svelte [hunk 3]
```diff
@@ -7,6 +7,7 @@ A container GraphicContext for the time series in a metrics dashboard.
   export let start: Date;
   export let end: Date;
   export let workspaceWidth: number;
+  export let enableFullWidth = false;
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-series/TimeSeriesChartContainer.svelte:10]: I don't see this prop overridden anywhere, but I assume it's for the upcoming feature?

File: web-common/src/features/dashboards/time-controls/TimeControls.svelte [hunk 4]
```diff
@@ -125,6 +129,21 @@
...
   }
 
+  function disableComparisons() {
+    metricsExplorerStore.displayComparison(metricViewName, false);
+    metricsExplorerStore.setComparisonDimension(metricViewName, undefined);
+  }
+
+  function enableComparison(type: string, name: string) {
+    if (type === "time") {
+      metricsExplorerStore.displayComparison(metricViewName, true);
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-controls/TimeControls.svelte:139]: Looks like this should now be `metricsExplorerStore.displayTimeComparison(...)`

File: web-common/src/features/dashboards/time-controls/TimeControls.svelte [hunk 5]
```diff
@@ -186,22 +205,31 @@
...
-      showComparison={$timeControlsStore?.showComparison}
-      selectedComparison={$timeControlsStore?.selectedComparisonTimeRange}
-      zone={$dashboardStore?.selectedTimezone}
-      comparisonOptions={availableComparisons}
+      on:disable-all-comparison={() => disableComparisons()}
+      showTimeComparison={$dashboardStore?.showComparison}
+      selectedDimension={$dashboardStore?.selectedComparisonDimension}
+      {dimensions}
     />
+    {#if $dashboardStore?.showComparison}
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-controls/TimeControls.svelte:217]: Likewise this should be `$dashboardStore?.showTimeComparison`

File: web-common/src/features/dashboards/time-series/multiple-dimension-queries.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  instanceId,
+  metricViewName,
+  dimensionName,
+  selectedMeasureNames,
+  filters,
+  start,
+  end,
+  interval,
+  zone
+) {
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-series/multiple-dimension-queries.ts:29]: It'd help to add types (inputs & output)

File: web-common/src/features/dashboards/time-series/MetricsTimeSeriesCharts.svelte
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-series/MetricsTimeSeriesCharts.svelte:1]: This file is pretty mammoth – at some point we should move the data fetching & transformation logic out into selectors

**Comment by ericpgreen2** [comment]
Left some small comments, but overall code looks great!

File: web-common/src/features/dashboards/time-series/multiple-dimension-queries.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
+// Throw away code for fetching timeseries data for individual dimension values
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-series/multiple-dimension-queries.ts:1]: "Throw away code"?

**Comment by ericpgreen2** [comment]
I'm not sure if this is existing behavior, but each dimension selection triggers two calls to `/totals`. One returns "count", one returns "total_records":
image
image

File: web-common/src/features/dashboards/big-number/MeasureBigNumber.svelte (hunk 1 above)
Comment (djbarnwal) [reply web-common/src/features/dashboards/big-number/MeasureBigNumber.svelte:45]: Yes, had `on:click` before but removed it. Have added these changes for the upcoming TDD feature.

File: web-common/src/features/dashboards/time-series/TimeSeriesChartContainer.svelte (hunk 3 above)
Comment (djbarnwal) [reply web-common/src/features/dashboards/time-series/TimeSeriesChartContainer.svelte:10]: Yes, it's for TDD view feature. Was prototyping it before dimension comparison. Didn't wanted to lose the additional code

File: web-common/src/features/dashboards/time-series/multiple-dimension-queries.ts (hunk 7 above)
Comment (djbarnwal) [reply web-common/src/features/dashboards/time-series/multiple-dimension-queries.ts]: We would be replacing this query with a single aggregation query on the backend in a couple of releases.

File: web-common/src/features/dashboards/time-series/MetricsTimeSeriesCharts.svelte
Comment (djbarnwal) [reply web-common/src/features/dashboards/time-series/MetricsTimeSeriesCharts.svelte:1]: Agreed. We need to move `totals` and `toplists` out of the component just as we did for time related stuff.

**Comment by djbarnwal** [comment]
> I'm not sure if this is existing behavior, but each dimension selection triggers two calls to `/totals`. One returns "count", one returns "total_records":

This is because we need the count for the Row Viewer at the bottom of the dashboard. We can just have one query and work it out as well but first we need to centralize our queries and put them out of the components.

File: web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte:1]: Just a note that this makes the `time-controls` surface/directory more of a general `controls` surface/directory.

File: web-common/src/features/dashboards/time-controls/TimeZoneSelector.svelte [hunk 8]
```diff
@@ -61,7 +61,7 @@
         }}
       >
         <div class="flex items-center gap-x-2">
-          <Globe />
+          <Globe size="16px" />
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-controls/TimeZoneSelector.svelte:64]: This triggers a browser console warning

File: web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    <MenuItem
+      focusOnMount={false}
+      selected={!(showTimeComparison || selectedDimension)}
+      on:before-select={() => {
+        intermediateSelection = NO_COMPARISON_LABEL;
+      }}
+      on:select={() => {
+        dispatch("disable-all-comparison");
+        toggleFloatingElement();
+      }}
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte:103]: This `on:before-select` and `intermediateSelection` looks over-complicated, but looks like you're just using the `MenuItem` component as designed?

**Review by ericpgreen2 — APPROVED** [review]
This is so awesome. I left one more UXQA concern in the Notion doc, but approving for you to merge when you wish.

File: web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte
Comment (djbarnwal) [reply web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte:1]: Janet mentioned that Brendan is going to work on restructuring this surface.

File: web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte (hunk 9 above)
Comment (djbarnwal) [reply web-common/src/features/dashboards/time-controls/ComparisonSelector.svelte:103]: Yes, it's pretty much the same as other menu components. IIRC this was done to change the state in UI without waiting for the state loop to come back in and change it. For slower dashboards, we wanted to show the change in selection as soon as possible without waiting for all the API calls to resolve.

**Pull request merged**